seismicpro/survey/headers.py

```diff
@@ -95,3 +96,117 @@ def callback(future, start_pos):
+    rec_cols = ["GroupX", "GroupY"]
+    cdp_cols = ["CDP_X", "CDP_Y"]
+    bin_cols = ["INLINE_3D", "CROSSLINE_3D"]
+
+    loaded_columns = headers.columns.values
+    available_columns = set(loaded_columns[headers.any(axis=0)])
+
+    zero_columns = set(loaded_columns) - set(available_columns)
+    if zero_columns:
+        msg_list.append("Empty headers: " + ", ".join(zero_columns))
+
+    if {"FieldRecord", "TraceNumber"} <= available_columns:
+        n_unique_ids = (~headers[["FieldRecord", "TraceNumber"]].duplicated()).sum()
+        if n_unique_ids != headers.shape[0]:
+            msg_list.append("Non-unique traces identifier (FieldRecord, TraceNumber)")
+
+    if {"FieldRecord", *shot_cols} <= available_columns:
+        fr_with_coords = headers[["FieldRecord", *shot_cols]].drop_duplicates()
+        n_unique_fr = fr_with_coords["FieldRecord"].nunique()
+        if len(fr_with_coords) != n_unique_fr:
+            msg_list.append("Several pairs of coordinates (SourceX, SourceY) for single FieldRecord")
+
+    if {'offset'} <= available_columns:
+        has_negative_offset = np.any(headers['offset'] < 0)
+        if has_negative_offset:
+            msg_list.append("Negative offset")
+
+    # Check that Euclidean distance calculated based on the coords from shot to receiver is close to the one stored
+    # in trace headers
+    if {*shot_cols, *rec_cols, "offset"} <= available_columns and not has_negative_offset:
+        real_offsets = headers["offset"].values
+        calculated_offsets = np.sqrt(np.sum((headers[shot_cols].values - headers[rec_cols].values)**2, axis=1))
+        has_mismatched_offsets = not np.allclose(real_offsets, calculated_offsets, rtol=offset_rtol, atol=offset_atol)
+        if has_mismatched_offsets:
+            msg_list.append("Mismatch of offsets in headers to the distance between shots (SourceX, "
+                            "\n    SourceY) and receivers (GroupX, GroupY) positions for each trace")
+
+    if {*cdp_cols, *bin_cols} <= available_columns:
+        unique_inline_cdp = headers[[*bin_cols, *cdp_cols]].drop_duplicates()
+        unique_cdp = unique_inline_cdp[cdp_cols].drop_duplicates()
+        if len(unique_inline_cdp) != len(unique_cdp):
+            msg_list.append("Non-unique mapping of geographic (CDP_X, CDP_Y) to line-based (INLINE_3D/"
+                            "\n    CROSSLINE_3D) coordinates")
+
+    if {*shot_cols, *rec_cols, *cdp_cols} <= available_columns:
+        raw_cdp = (headers[shot_cols].values + headers[rec_cols].values) / 2
+        if not np.allclose(raw_cdp, headers[cdp_cols].values, rtol=0, atol=cdp_atol):
+            if has_mismatched_offsets:
+                msg = "Inconsistent range of some geographic coordinates (SourceX, SourceY, GroupX,"\
+                      "\n    GroupY, CDP_X, CDP_Y)"
+            else:
+                msg = "Inconsistent range of CDP_X and CDP_Y coordinates compare to SourceX, "\
+                      "\n    SourceY, GroupX, GroupY"
```

Comment:
```suggestion
                msg = "Inconsistent range of CDP_X and CDP_Y coordinates compared to SourceX, "\
                      "\n    SourceY, GroupX, GroupY"
```
Why not keep only this message?